### PR TITLE
kubectl-explain: add --output plaintext-openapiv2 fallback

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

- Adds option to `--output` argument `plaintext-openapiv2` which refers to old implementation. 
- Refactors unit tests to try both OpenAPI versions
- Adds test for fallback
- 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Adds --output plaintext-openapiv2 argument to kubectl explain to use old openapiv2 `explain` implementation.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP] 3515 OpenAPI v3 for kubectl explain: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/3515-kubectl-explain-openapiv3
```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/util/factory.go [hunk 1]
```diff
@@ -63,4 +64,7 @@ type Factory interface {
 	Validator(validationDirective string) (validation.Schema, error)
 	// OpenAPISchema returns the parsed openapi schema definition
 	OpenAPISchema() (openapi.Resources, error)
+	// OpenAPIV3Schema returns a client for fetching parsed schemas for
+	// any group version
+	OpenAPIV3Client() (openapiclient.Client, error)
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/util/factory.go:69]: I think, one day(out of scope of this PR), we need to cut out OpenAPI functions from Factory and instead move them to in new `OpenAPIFactory`. 

Because for example, isn't this new function only used for `kubectl explain`?. By adding this in Factory, we are passing this function into all commands.

Comment (apelisse) [reply staging/src/k8s.io/kubectl/pkg/cmd/util/factory.go:69]: `OpenAPISchema()` is used throughout, `OpenAPIV3Client()` is used only in explain, but that's just because we haven't wired it further yet. @seans3 is working on replacing some of the `OpenAPISChema()` with `OpenAPIV3Client()` in various places. 

Now that being said, I think the mega factory object is inconvenient to mock in tests, but I don't know how I would refactor that.

Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/util/factory.go:69]: `Validator`, `OpenAPISchema` and `OpenAPIV3Client` functions can be in `OpenAPIFactory`. All commands might use it something like that;

```go
func NewCmdDiff(f cmdutil.Factory, of cmdutil.OpenAPIFactory, streams genericclioptions.IOStreams) *cobra.Command {
}
```

In my opinion, that will alleviate the responsibilities of mega Factory. If it looks like a good idea, I can propose a PR after @alexzielenski  and @seans3's changes are done.

Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/util/factory.go:69]: ah, there is already a TODO for it https://github.com/kubernetes/kubernetes/blob/1da46b2f28267421dad9a8e5164e6bce16ae9fcb/staging/src/k8s.io/kubectl/pkg/cmd/util/factory.go#L38 :)

**Review by apelisse — COMMENTED** [review]
I haven't looked at the test, it looked very complicated.

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go [hunk 2]
```diff
@@ -172,16 +172,53 @@ func (o *ExplainOptions) Run() error {
...
-			o.DiscoveryClient.OpenAPIV3(),
-			fullySpecifiedGVR,
-			recursive,
-			o.OutputFormat,
-		)
+		format := o.OutputFormat
+
+		if format == defaultTemplateName {
+			// Default to "plaintext"
+			format = plaintextTemplateName
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:181]: saying "plaintext" directly implies no fallback mechanism. The default is to try v3 then v2. If you say specifically what you want then it does not try anything else.

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go [hunk 3]
```diff
@@ -172,16 +172,53 @@ func (o *ExplainOptions) Run() error {
 		}
 	}
 
+	// Fallback to openapiv2 implementation using special template name
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:175]: I will open another PR that removes this conditional and will add/update tests accordingly

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go [hunk 4]
```diff
@@ -74,15 +78,15 @@ type ExplainOptions struct {
 	OutputFormat string
 
 	// Client capable of fetching openapi documents from the user's cluster
-	DiscoveryClient discovery.DiscoveryInterface
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:77]: not sure how this comment relates to the attached line of code

[the remark this answers is not in the document]
Comment (apelisse) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:77]: It doesn't, I can't comment on the specific line :-)

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go]: changed defaulting as discussed offline

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go [hunk 5]
```diff
@@ -172,16 +171,49 @@ func (o *ExplainOptions) Run() error {
...
-			fieldsPath,
-			o.Out,
-			o.DiscoveryClient.OpenAPIV3(),
-			fullySpecifiedGVR,
-			recursive,
-			o.OutputFormat,
-		)
+		format := o.OutputFormat
+
+		if format == plaintextTemplateName {
```
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:178]: WDYT of making this into a switch statement, maybe with a fallback if v3 isn't supported?

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go [hunk 6]
```diff
@@ -96,60 +104,193 @@ func TestExplainNotExistResource(t *testing.T) {
...
 }
 
-func TestExplainNotExistVersion(t *testing.T) {
-	tf := cmdtesting.NewTestFactory()
-	defer tf.Cleanup()
+type explainTestCase struct {
+	Name               string
+	Args               []string
+	Flags              map[string]string
+	Env                map[string]string
```
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:111]: I was going to say that you should probably not use this, but then I realized it's not used.

Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go]: removed

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go (hunk 5 above)
Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go]: changed to switch stmt

**Comment by apelisse** [comment]
I've looked at it, I could spend more time with it but it generally looks good, do you want to LGTM @ardaguclu ?

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go [hunk 7]
```diff
@@ -127,11 +130,10 @@ func (o *ExplainOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args []
 
 	// Only openapi v3 needs the discovery client.
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:131]: nit:

```suggestion
	// Only openapi v3 needs the openapiv3client.
```

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go [hunk 8]
```diff
@@ -14,24 +14,32 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package explain
+package explain_test
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:17]: What is the reason of this change? this makes `explain` command tests unstandard?

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go [hunk 9]
```diff
@@ -96,60 +104,188 @@ func TestExplainNotExistResource(t *testing.T) {
...
+		},
 	}
 
-	cmd.Flags().Set("api-version", "batch/v1beta1")
-	cmd.Run(cmd, []string{"cronjobs"})
-	if !strings.Contains(buf.String(), "VERSION:  batch/v1beta1") {
-		t.Fatalf("expected output should include pod batch/v1beta1")
+	type catchFatal error
+
+	for _, vers := range []string{openapiv2Name, openapiv3Name} {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:212]: Instead boiling openapiv2 and openapiv3 tests down to the one function, I'd suggest to run them in their own functions. For example, when openapiv2 is deprecated, this function will be changed fundamentally again(which hurts confidence due to making changes both in code and unit tests.). Moreover, there won't be any need to this loop and `VersionWhiteList` field.

**Review by ardaguclu — COMMENTED** [review]
Code looks good to me. Just left a few comments for unit tests.

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go (hunk 9 above)
Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go]: good point. refactored into separate tests and removed field

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go (hunk 8 above)
Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:17]: im not sure what you mean by unstandard, but using _test package with go makes them white box tests, which IMO are preferable to blackbox test package if we can

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go (hunk 7 above)
Comment (alexzielenski) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go]: fixed

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go (hunk 8 above)
Comment (apelisse) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:17]: yeah, it means you can't depend on internal, which is definitely better.

**Review by BenTheElder — COMMENTED** [review]
/approve
for vendor/

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go (hunk 8 above)
Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:17]: I mean by "unstandard" comparing to the other commands unit tests which all of them are whitebox tests. But I agree with you.
